Reactive utilities

romain-grecourt opened the pull request:
- Add new reactive utilities: Single and Multi to replace usage of reactor.Flux and Reactor.Mono.
- Minor refactoring of OriginThreadPublisher to make it re-usable and move it under common/reactive.
- Improved javadocs around the ReferenceHoldingQueue mechanism

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/Multi.java:
```diff
@@ -0,0 +1,221 @@
+     *
+     * @return Mono
+     */
+    default Mono<String> collectString() {
```
This is not nice

romain-grecourt replied:
Removed.

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/OriginThreadPublisher.java:
```diff
@@ -24,107 +24,115 @@
+     * data; to not send when the number of permits is zero).
      */
-    OriginThreadPublisher(UnboundedSemaphore semaphore, ReferenceHoldingQueue<ByteBufRequestChunk> referenceQueue) {
+    public OriginThreadPublisher(UnboundedSemaphore semaphore) {
```
Please change to `protected`

romain-grecourt replied:
Changed it to protected.

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/OriginThreadPublisher.java:
```diff
@@ -24,107 +24,115 @@
      */
-    OriginThreadPublisher(ReferenceHoldingQueue<ByteBufRequestChunk> referenceQueue) {
-        this(new UnboundedSemaphore(), referenceQueue);
+    public OriginThreadPublisher() {
```
`protected`

romain-grecourt replied:
Changed it to protected.

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/OriginThreadPublisher.java:
```diff
@@ -24,107 +24,115 @@
                 public void request(long n) {
                     if (n <= 0) {
-                        error(new IllegalArgumentException("[3.9] Illegal value requested: " + n));
+                        error(new IllegalArgumentException(
```
Please do not line break

romain-grecourt replied:
I fixed a lot of the over re-formats, including this one.

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/OriginThreadPublisher.java:
```diff
@@ -165,76 +173,63 @@ public void cancel() {
-     */
-    void hookOnCancel() {
-    }
+    protected abstract T wrap(U data);
```
Maybe have a default case `return (T) data`?

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/UnboundedSemaphore.java:
```diff
@@ -0,0 +1,87 @@
+ */
+
+package io.helidon.common.reactive;
+
```
Probably forgot to remove the original class in webserver, so marked as new.

romain-grecourt replied:
Removed the original.

tomas-langer commented on media/common/src/main/java/io/helidon/media/common/CharsetCache.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Utility for caching per commonly used charsets.
+ */
+final class CharsetCache<T> {
```
This class is not used within this PR, please remove

romain-grecourt replied:
Removed it.

tomas-langer commented on media/common/src/main/java/io/helidon/media/common/ContentReaders.java:
```diff
@@ -19,136 +19,139 @@
      */
-    static Reader<String> cachedStringReader(String charset) {
-        return CACHED_READERS.get(charset);
+    public static io.helidon.common.http.Reader<byte[]> byteArrayReader() {
```
Use import instead of FQDN

romain-grecourt replied:
Done.

tomas-langer commented on webserver/webserver/src/main/java/io/helidon/webserver/ByteBufRequestChunk.java:
```diff
@@ -53,7 +57,8 @@ public boolean isReleased() {
     public ByteBuffer data() {
         if (isReleased()) {
-            throw new IllegalStateException("The request chunk was already released!");
+            throw new IllegalStateException(
```
line length...

romain-grecourt replied:
Fixed the re-formatting.

tomas-langer commented on webserver/webserver/src/main/java/io/helidon/webserver/ByteBufRequestChunk.java:
```diff
@@ -83,24 +89,41 @@ protected void finalize() {
-                                   + "explicitly when they're not needed.");
+            LOGGER.warning("LEAK: DataChunk.release() was not called before it was garbage collected. "
+                    + "While the Reactive WebServer is "
+                    + "designed to automatically release all the RequestChunks, it still "
```
RequestChunks -> DataChunks

romain-grecourt replied:
Fixed.

tomas-langer commented on common/reactive/src/test/java/io/helidon/common/reactive/valve/ValvePublisherTest.java:
```diff
@@ -38,164 +39,172 @@
     }
 
-    @Test
+    //@Test
```
Commented test

romain-grecourt replied:
Removed commented out tests.

romain-grecourt commented on common/reactive/src/main/java/io/helidon/common/reactive/Mono.java:
```diff
@@ -0,0 +1,202 @@
+     * @throws IllegalStateException if the Mono wraps an error or if
+     * interrupted
+     */
+    default T block() {
```
Removed these methods.

romain-grecourt commented on common/reactive/src/main/java/io/helidon/common/reactive/Mono.java:
```diff
@@ -0,0 +1,202 @@
+/**
+ * Single item publisher.
+ * @param <T> published type
+ */
```
Renamed it to Single.

romain-grecourt commented on common/reactive/src/main/java/io/helidon/common/reactive/Mono.java:
```diff
@@ -0,0 +1,202 @@
+     * @param mapper mapper
+     * @return Mono
+     */
+    default <U> Mono<U> map(MonoMapper<T, U> mapper) {
```
I've made all the processors package private and exposed interfaces instead.
I've created my own Mapper interface for now since the common-mapper is not available yet ; we need to get a PR going for that separately out of the dbclient branch.

romain-grecourt commented on common/reactive/src/main/java/io/helidon/common/reactive/Mono.java:
```diff
@@ -0,0 +1,202 @@
+     * @return CompletableFuture
+     */
+    default CompletableFuture<T> toFuture() {
+        try {
```
The reason you suggested using `Optional<T>` was because the current implementation completes the future with `null` when the `Single`/`Mono` completes without value. Instead we can complete the future exceptionally with an exception.

`CompletableFuture<Optional<T>>` feels too much impractical for most use-cases.

romain-grecourt commented on common/reactive/src/main/java/io/helidon/common/reactive/Mono.java:
```diff
@@ -0,0 +1,202 @@
+     * @return Mono
+     */
+    @SuppressWarnings("unchecked")
+    static <T> Mono<T> from(Publisher<? extends T> source) {
```
I think it's okay to change the current `from` implementation to expect a publisher of only one item. However I wouldn't add a method `firstOf` on `Single`, instead having a `Multi.first()` method seems more appropriate.

romain-grecourt commented:
@tomas-langer I think I've addressed all your review comments. I created my own mapper interface since the common-mapping module is not available yet.

Can the mapper-module be extracted from the dbclient branch and added separately ? Is it ready to go ?

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/Collector.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Get the collected items container.
+     *
+     * @return T
```
typo? `@return T` should probably be `@return U`

romain-grecourt replied:
Good catch, will fix.

romain-grecourt replied:
Fixed.

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/Single.java:
```diff
@@ -0,0 +1,142 @@
+     *
+     * @return CompletableFuture
+     */
+    default CompletableFuture<T> toFuture() {
```
Shouldn't this return a `CompletionStage` rather than a `CompletableFuture`? Future can be explicitly completed, which probably would not work well with the `Single` instance.

romain-grecourt replied:
Right, the main concern is that this is going to be only used to block ; if we were to change this to a `CompletionStage` than instead of `single.toFuture().get()` you'd have to do `single.toFuture().toCompletableFuture().get()`.

Is that okay ? Maybe we could extend `CompletableFuture` and override the complete the methods to throw an `UnsupportedOperationException` ?

romain-grecourt replied:
We could also expose an interface like `interface SingleFuture extends Future<T>, CompletionStage<T> {}` and return a CompletableFuture as the instance.

tomas-langer replied:
If you think this is only useful for calling `future.get()`, then you can maybe return `Future<T>`?
If we expect people to use `thenApply` etc., then it should be `CompletionStage` and you would need to call the `toCompletableFuture()` on it. 
I would rather not create a new interface to combine `Future` and `CompletionStage` - last possible solution imho

romain-grecourt replied:
Roger that. Returning a future sounds good.

romain-grecourt replied:
I've changed the method to return a `CompletionStage` and added two variants of `.get` as short hands for `single.toFuture().toCompletableFuture().get()`.

I've also updated the implementation class `SingleToFuture` to override `complete()` and `completeException` to throw an `UnsupportedOperationException` to prevent manual completion.

tomas-langer commented on common/reactive/src/main/java/io/helidon/common/reactive/SubmissionPublisher.java:
```diff
@@ -30,13 +30,13 @@
+ * A {@link Flow.Publisher} that asynchronously issues submitted (non-null) items to current subscribers until it is closed.
  *
  * @param <T> the published item type
+ * @deprecated This class will be removed in the next major release.
```
Please explain what to use instead/what is used instead.
Also please mark class as `@Deprecated`

romain-grecourt replied:
I'm not planning on a direct replacement, `OriginThreadPublisher` is the closest thing.
Will add `@Deprecated`.

romain-grecourt replied:
Fixed.

tomas-langer commented on common/reactive/src/test/java/io/helidon/common/reactive/WrappedOriginThreadPublisher.java:
```diff
@@ -26,15 +23,15 @@
 /**
  * The WrappedOriginThreadPublisher.
```
Class is public - the javadoc is not really describing what it is.

romain-grecourt replied:
This class is under `src/test`, it's used to run the reactive-stream TCK.

tomas-langer replied:
oops :)
